notebooks/community/aisamples/AIsample - Uplift Modelling.ipynb

```diff
@@ -0,0 +1,882 @@
+    "    ranked[\"label\"] = uplift_df[label_col]\n",
+    "    ranked[\"uplift_score\"] = uplift_df[uplift_col]\n",
+    "    ranked[\"ranked_uplift\"] = ranked.uplift_score.rank(pct=True, ascending=False)\n",
+    "    ranked = ranked.sort_values(by=\"ranked_uplift\").reset_index(drop=True)\n",
+    "    return ranked\n",
+    "\n",
+    "\n",
+    "def uplift_eval(ranked):\n",
+    "    uplift_df = ranked.copy()\n",
+    "    # Using Treatment and Control Group to calculate the uplift (Incremental gain)\n",
+    "    C, T = sum(ranked.treatment == 0), sum(ranked.treatment != 0)\n",
+    "    ranked[\"cr\"] = ranked.label\n",
+    "    ranked[\"tr\"] = ranked.label\n",
+    "    ranked.loc[ranked.treatment != 0, \"cr\"] = 0\n",
+    "    ranked.loc[ranked.treatment == 0, \"tr\"] = 0\n",
+    "    ranked[\"cr/c\"] = ranked.cr.cumsum() / C\n",
+    "    ranked[\"tr/t\"] = ranked.tr.cumsum() / T\n",
+    "    # Calculate and put the uplift value into dataframe\n",
+    "    uplift_df[\"uplift\"] = round(ranked[\"tr/t\"] - ranked[\"cr/c\"], 5)\n",
+    "\n",
+    "    # Add q0\n",
+    "    q0 = pd.DataFrame({\"ranked_uplift\": 0, \"uplift\": 0, \"treatment\": None}, index=[0])\n",
+    "    uplift_df = pd.concat([q0, uplift_df]).reset_index(drop=True)\n",
+    "\n",
+    "    return uplift_df\n",
+    "\n",
+    "\n",
+    "def uplift_plot(uplift_df):\n",
+    "    gain_x = uplift_df.ranked_uplift\n",
+    "    gain_y = uplift_df.uplift\n",
+    "    # plot the data\n",
+    "    plt.figure(figsize=(10, 6))\n",
+    "    mpl.rcParams[\"font.size\"] = 8\n",
+    "\n",
+    "    ax = plt.plot(gain_x, gain_y, color=\"#2077B4\", label=\"Normalized Uplift Model\")\n",
+    "\n",
+    "    plt.plot(\n",
+    "        [0, gain_x.max()],\n",
+    "        [0, gain_y.max()],\n",
+    "        \"--\",\n",
+    "        color=\"tab:orange\",\n",
+    "        label=\"Random Treatment\",\n",
+    "    )\n",
+    "    plt.legend()\n",
+    "    plt.xlabel(\"Porportion Targeted\")\n",
+    "    plt.ylabel(\"Uplift\")\n",
+    "    plt.grid(b=True, which=\"major\")\n",
+    "\n",
+    "    return ax"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "Now we can plot the uplift curve on the prediction of the test dataset"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "ranked_df = uplift_rank(\n",
+    "    test_pred_pandas_df,\n",
+    "    treatment_col=TREATMENT_COLUMN,\n",
+    "    label_col=LABEL_COLUMN,\n",
+    "    uplift_col=\"lift_pred\",\n",
+    ")\n",
+    "uplift_df = uplift_eval(ranked_df)\n",
+    "uplift_plot(uplift_df)"
```

Comment:
This is your WOW moment here, consider puting a pucture of this plot in the notebook, Feel free to upload an image to https://mmlspark.blob.core.windows.net/graphics/notebooks/ to achieve this 

If you are really feeling ambitious, make this in plotly and add some interesting info ont he hover tips, it will be like an interactive dashboard and very exciting to showcase